src/modules/fw_pos_control/FixedwingPositionControl.cpp

```diff
@@ -2840,47 +2813,97 @@ void FixedwingPositionControl::publishOrbitStatus(const position_setpoint_s pos_
+		// is missed for any reason. in the future this logic should all be handled in one place in a dedicated
+		// flight mode state machine.
+		navigateWaypoint(end_waypoint, vehicle_pos, ground_vel, wind_vel);
+		return;
+	}
+
+	// follow the line segment between the start and end waypoints
+	navigateLine(start_waypoint, end_waypoint, vehicle_pos, ground_vel, wind_vel);
+}
+
+void FixedwingPositionControl::navigateWaypoint(const Vector2f &waypoint_pos, const Vector2f &vehicle_pos,
+		const Vector2f &ground_vel, const Vector2f &wind_vel)
+{
+	const Vector2f vehicle_to_waypoint = waypoint_pos - vehicle_pos;
+
+	if (vehicle_to_waypoint.norm() < FLT_EPSILON) {
+		// degenerate case: the vehicle is on top of the single waypoint. (can happen). maintain the last npfg command.
+		return;
+	}
+
+	const Vector2f unit_path_tangent = vehicle_to_waypoint.normalized();
+	_closest_point_on_path = waypoint_pos;
+
+	const float path_curvature = 0.f;
+	_npfg.guideToPath(vehicle_pos, ground_vel, wind_vel, unit_path_tangent, _closest_point_on_path, path_curvature);
+
+	// for logging - note we are abusing path tangent vs bearing definitions here. npfg interfaces need to be refined.
+	_target_bearing = atan2f(unit_path_tangent(1), unit_path_tangent(0));
+}
+
+void FixedwingPositionControl::navigateLine(const Vector2f &point_on_line_1, const Vector2f &point_on_line_2,
+		const Vector2f &vehicle_pos, const Vector2f &ground_vel, const Vector2f &wind_vel)
+{
+	const Vector2f line_segment = point_on_line_2 - point_on_line_1;
 
-	} else if (vector_B_to_A.dot(vector_B_to_vehicle) < -FLT_EPSILON) {
-		// we are behind waypoint B, fly directly to it
-		desired_path = -vector_B_to_vehicle;
+	if (line_segment.norm() <= FLT_EPSILON) {
+		// degenerate case: line segment has zero length. maintain the last npfg command.
+		return;
 	}
 
-	// track the line segment
-	Vector2f unit_path_tangent{desired_path.normalized()};
+	const Vector2f unit_path_tangent = line_segment.normalized();
+
+	const Vector2f point_1_to_vehicle = vehicle_pos - point_on_line_1;
+	_closest_point_on_path = point_on_line_1 + point_1_to_vehicle.dot(unit_path_tangent) * unit_path_tangent;
+
+	const float path_curvature = 0.f;
+	_npfg.guideToPath(vehicle_pos, ground_vel, wind_vel, unit_path_tangent, _closest_point_on_path, path_curvature);
+
+	// for logging - note we are abusing path tangent vs bearing definitions here. npfg interfaces need to be refined.
 	_target_bearing = atan2f(unit_path_tangent(1), unit_path_tangent(0));
-	_closest_point_on_path = waypoint_A + vector_A_to_vehicle.dot(unit_path_tangent) * unit_path_tangent;
-	_npfg.guideToPath(vehicle_pos, ground_vel, wind_vel, unit_path_tangent, waypoint_A, 0.0f);
-} // navigateWaypoints
+}
+
+void FixedwingPositionControl::navigateLine(const Vector2f &point_on_line, const float line_bearing,
+		const Vector2f &vehicle_pos, const Vector2f &ground_vel, const Vector2f &wind_vel)
```

Comment:
This function is very similar to navigatePathTangent. I would propose to simply call here the navigatePathTangent function and move the stuff from here to there (here is the better closest point on path calculations as the other one simply uses the input point). Also here we need to think if we want to support both handing in either the bearing angle and the tangent vector or what we want to do since internally we will anyways also calculate the other representation as well for logging (or rethink the problem of needing the path tangent for the npfg function and the angle for logging).

Reply:
> This function is very similar to navigatePathTangent.

it is. but i would argue the navigatePathTangent is the one I don't like. it was more just a quick hack to enable @Jaeyoung-Lim for his velocity control. Once we rework npfg API (see comment below) - this becomes a direct call to NPFG for @Jaeyoung-Lim instead of this strange function. I would propose to leave it alone for now.. and make the proper refactoring for 1.15.

> or rethink the problem of needing the path tangent for the npfg function 

yeah i want to change the interface. as we discussed a while back -- course control (and airspeed control separated in parallel). but this requires a bit more refactoring than just a patch for 1.14 that needs to get out soon.

Reply:
My intention of navigatePathTangent was not a hack, but to unify "navigate" functions into a single interface. However, this got lost during https://github.com/PX4/PX4-Autopilot/pull/18980, where I tried to separate the path handling and the waypoint handling. IMHO, the guidance controller should "only" receive information about the reference or combinations of references (e.g. position / position + tangent / position + tangent + curvature).

This gives us a good reference to design an interface that is solely on reference for the NPFG(basically what guide to path is doing). I think the discussion here is more about the waypoint handling, which I wish we could take out of fw_pos_control completely anyway.

Reply:
> waypoint handling, which I wish we could take out of fw_pos_control completely

Yes - waypoint handling should be a set of library methods that a fixed-wing flight mode can use to generate paths (path segments themselves as setpoints), or path following setpoints (a set of state setpoints that steer the vehicle to the path, generally course and acceleration). Paths should be the real interface we strive for as input to path following controller. But the reason this whole path tangent discussion comes up is just that we dont have that ideal interface currently.

Ideal case is:
path -> [path follower] -> path following setpoints -> [course controller] -> course following setpoints 

without path api:
path -> [path handler (find closest point to vehicle on path)] -> instantaneous path setpoint (point on path, bearing, curvature) -> [path follower] -> path following setpoints -> [course controller] -> course following setpoints

The reason to keep the instantaneous path setpoint interface is only for the case that we externally handle the paths. But in this case the path handler computes half the calculations that the path follower normally should, then the path follower ends up repeating those same calculations.

Why keep these definitions? Because it clearly delineates what we are actually wanting to command from the limits of our current interfaces. If I want to directly command a course setpoint, e.g. from a ROS mode, it is IMO an abuse to need the vehicle's current position to input as instantaneous path setpoint at some higher latency from external link which is then evaluated potentially at lower latency repeatedly on the microcontroller repeating that crosstrack line error calculation. I should just be able to command course without any knowledge of the vehicle's position and let the lower level loops track that setpoint. Hence.. for now, since we don't have a path interface, sure, we can keep instantaneous path setpoint interface, but it should not be unified for all input methods.

This is anyway a deeper discussion we should have over fixed-wing architecture / refactor for 1.15. For now.. we just need to make sure we aren't letting planes do dangerous things on 1.14.

Reply:
>  I should just be able to command course without any knowledge of the vehicle's position and let the lower level loops track that setpoint.

Okay, I definitely agree with this point and was not implying the interface should look this way.